File: backend/apps/marketplace/installs.py
"""What this machine installed from the marketplace, by listing id.

The App Store knows what is installed; without a record every package showed Install again after a
restart, and Open had nothing to open. The importer creates a fresh entity per commit and returns its
id, so the record is written at commit time and maps a listing to the thing it became. A record whose
entity has since been deleted is stale by design: the store re-checks the live lists before showing Open.
"""

import json
import os
import time
from typing import Dict, Optional

from pydantic import BaseModel, ConfigDict
from typeguard import typechecked

from backend.config.paths import DATA_ROOT

INSTALLS_PATH = os.path.join(DATA_ROOT, "marketplace_installs.json")


class InstallRecord(BaseModel):
    """One pointer per importable root, typed, so an id can never point at a kind nobody can resolve."""

    model_config = ConfigDict(validate_assignment=True)

    listing_id: str
    root_type: str
    output_id: Optional[str] = None
    skill_id: Optional[str] = None
    workflow_id: Optional[str] = None
    dashboard_id: Optional[str] = None
    session_id: Optional[str] = None
    version: str = ""
    installed_at: float = 0.0

    def root_id(self) -> Optional[str]:
        return self.output_id or self.skill_id or self.workflow_id or self.dashboard_id or self.session_id


@typechecked
def load_installs(path: Optional[str] = None) -> Dict[str, InstallRecord]:
    p = path or INSTALLS_PATH
    try:
        with open(p, "r", encoding="utf-8") as f:
            raw = json.load(f)
    except (FileNotFoundError, json.JSONDecodeError, OSError):
        return {}
    out: Dict[str, InstallRecord] = {}
    for key, value in (raw or {}).items():
        try:
            out[key] = InstallRecord.model_validate(value)
        except Exception:
            continue
    return out


@typechecked
def record_install(rec: InstallRecord, path: Optional[str] = None) -> Dict[str, InstallRecord]:
    p = path or INSTALLS_PATH
    installs = load_installs(p)
    if rec.installed_at == 0.0:
        rec = rec.model_copy(update={"installed_at": time.time()})
    installs[rec.listing_id] = rec
    os.makedirs(os.path.dirname(p) or ".", exist_ok=True)
    tmp = p + ".tmp"
    with open(tmp, "w", encoding="utf-8") as f:
        json.dump({k: v.model_dump() for k, v in installs.items()}, f, indent=1)
    os.replace(tmp, p)
    return installs

File: frontend/src/app/pages/Directory/packages/installs.ts
import { API_BASE } from '@/shared/config';
import type { Listing } from './catalog';

// One typed pointer per importable root, mirroring the backend record.
export interface InstallRecord {
  listing_id: string;
  root_type: string;
  output_id?: string | null;
  skill_id?: string | null;
  workflow_id?: string | null;
  dashboard_id?: string | null;
  session_id?: string | null;
  version: string;
  installed_at: number;
}

export function recordFor(listingId: string, rootType: string, rootId: string, version: string): Omit<InstallRecord, 'installed_at'> {
  const rec: Omit<InstallRecord, 'installed_at'> = { listing_id: listingId, root_type: rootType, version };
  if (rootType === 'app') rec.output_id = rootId;
  else if (rootType === 'skill') rec.skill_id = rootId;
  else if (rootType === 'workflow') rec.workflow_id = rootId;
  else if (rootType === 'dashboard') rec.dashboard_id = rootId;
  else if (rootType === 'session' || rootType === 'mode') rec.session_id = rootId;
  return rec;
}

export type PillState = 'get' | 'installing' | 'open' | 'installed';

interface LiveEntities {
  outputs: Record<string, unknown>;
  skills: Record<string, unknown>;
  workflows: Record<string, unknown>;
}

// Get until the record says otherwise; Open only while the thing it became still exists, so a deleted app offers Get again instead of an Open that goes nowhere.
export function installState(listing: Listing, record: InstallRecord | undefined, live: LiveEntities): PillState {
  if (!record) return 'get';
  if (record.output_id) return record.output_id in live.outputs ? 'open' : 'get';
  if (record.skill_id) return record.skill_id in live.skills ? 'open' : 'get';
  if (record.workflow_id) return record.workflow_id in live.workflows ? 'open' : 'get';
  return 'installed';
}

export async function fetchInstalls(): Promise<Record<string, InstallRecord>> {
  const res = await fetch(`${API_BASE}/marketplace/installed`);
  if (!res.ok) return {};
  const body = (await res.json()) as { installs: Record<string, InstallRecord> };
  return body.installs || {};
}

export async function recordInstall(rec: Omit<InstallRecord, 'installed_at'>): Promise<Record<string, InstallRecord>> {
  const res = await fetch(`${API_BASE}/marketplace/installed`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ ...rec, installed_at: 0 }),
  });
  if (!res.ok) throw new Error('Installed, but it may show Install again after a restart.');
  const body = (await res.json()) as { installs: Record<string, InstallRecord> };
  return body.installs || {};
}
